feat: 월간 인기글 랭킹 Top 3 수상 기능 구현 [ISSUE-324]

yelimkim98 opened the pull request:
**이슈 번호**

resolved: #324
resolved: #504

**작업 내용**
-------------------
### 1 월간 인기글 랭킹 Top 3 수상 기능 구현
 - 페이징을 사용하지 않는 Posts 조회, Ranking 조회 구현
 - 인기글 수상 서비스, 상 이름 maker 구현  
  
### 2 RankingCriteria 지난달, 지난주, 어제 계산 구현 잘못된 부분 수정  

### 3 테스트 코드 보강  

### 4 랭킹 프론트 동순위 로직 제거
-------------------
**테스트 작성 여부**

- [X] Test case

**주의 사항**

yelimkim98 committed:
feat: 인기글 수상 스케줄러 클래스 및 메서드 뼈대 구현

yelimkim98 committed:
feat: PostRepository 페이징 없이 랭킹 가져오기 구현

yelimkim98 committed:
feat: RankingService 페이징 없이 랭킹 가져오기 구현

yelimkim98 committed:
refactor: 메서드 네이밍 수정

yelimkim98 committed:
feat: 전체 지역 가져오기 기능 구현

yelimkim98 committed:
refactor: 부문 창시 상 만들기를 sector service 에서 award service 로 옮김

yelimkim98 committed:
fix: RankingCriteria 시작시각 및 끝시각 로직 수정 및 테스트코드 작성

yelimkim98 committed:
feat: 매월 1일마다 지난달 기록으로 인기글 상을 수여하는 기능 구현

yelimkim98 committed:
Test: RankingCriteriaTest 구현

yelimkim98 committed:
test: 이미지 응답의 시간 정보 소수점 자릿수때문에 생기는 test fail 해결

yelimkim98 committed:
fix: 인기글 상 수여 기능 리팩토링 및 테스트 & 이를 통해 발견한 문제 해결

yelimkim98 committed:
refactor: 객체 책임 분리, 네이밍, 컨벤션 개선

yelimkim98 committed:
refactor: 수상 기준 수정 및 상수화

yelimkim98 committed:
fix: 프론트 랭킹 페이지 동순위 로직 제거

yelimkim98 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#324

yelimkim98 committed:
fix: 백엔드 401 응답이 생겼으므로 프론트 401 login redirect 처리

yelimkim98 committed:
refactor: 로그아웃 버튼 커서모양 변경

yelimkim98 committed:
refactor: 의미없는 파일체인지 원상복구

쓸데없이 풀리퀘에 파일 체인지가 많아서 불편했습니다.

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/util/AwardNameMaker.java:
```diff
@@ -12,4 +16,88 @@
+        String year = Integer.toString(startDate.getYear());
+        String month = Integer.toString(startDate.getMonthValue());
+
+        return year
+            + "년 "
+            + month
+            + "월 "
+            + targetArea.getLastDepthName()
+            + " "
+            + popularPost.getSector().getName()
+            + " 부문 "
+            + ranking
+            + "위";
```
이 부분은 `+` 연산자가 많아서 `StringBuilder`를 써도 좋겠네요 :)

yelimkim98 replied:
이부분 원래 StringBuilder를 써서 구현했었는데요..! 인텔리제이가 String 으로 할것을 추천하더라구요..?
![image](https://user-images.githubusercontent.com/28949340/95558541-5636e900-0a51-11eb-81a8-5c61b9f0a84d.png)


그래서 String 과 StringBuilder 를 사용할 때 차이를 찾아봤는데
명확한 해답은 못찾았지만
지금 상황에서는 큰 차이가 없을 것 같다고 생각되더라구요..!

참고한 링크
1
(번역)
IntelliJ의 제안 `'StringBuilder' can be replaced with 'String'` 은 결과 연결이 최소한 원래 코드만큼 효율적이거나 원래 코드보다 효율적인 경우에만 보고합니다.

(Inspection 'StringBuilder' can be replaced with 'String' only reports the cases when the resulting concatenation is at least as efficient or more efficient than the original code.)

https://intellij-support.jetbrains.com/hc/en-us/community/posts/115000093250-Convert-StringBuilder-to-String-concatenation-suggestion

2
작은양의 "+" 연산은 오히려 StringBuilder를 사용하는 것이 안좋다고 한다.

https://m.blog.naver.com/PostView.nhn?blogId=bluelevel&logNo=70130739583&proxyReferer=https:%2F%2Fwww.google.com%2F

aegis1920 replied:
적은 양의 "+" 연산이 아닌 것 같아서 이야기했었는데 링크해주신 곳에서 답변을 보니 반복문을 쓰면 StringBuilder를 사용하고 아니면 String을 쓴다고 하네요 :)
깨달음 주셔서 감사합니다 👍 👍

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/util/AwardNameMaker.java:
```diff
@@ -12,4 +16,88 @@
+            return;
+        }
+        throw new IllegalArgumentException(
+            "start date 와 end date 의 연도가 잘못되었습니다.\n"
+                + "start date : "
+                + startDate.getMonthValue() + " 월 "
+                + startDate.getDayOfMonth() + "일\n"
+                + "end date : "
+                + endDate.getMonthValue() + " 월 "
+                + endDate.getDayOfMonth() + "일"
```
여기도 마찬가지 StringBuilder!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/AwardController.java:
```diff
@@ -12,25 +12,25 @@
 public class AwardController {
 
-    private final AwardService awardService;
+    private final AwardReadService awardReadService;
```
`AwardService`가 클래스명에 더 맞지 않을까요?

yelimkim98 replied:
`AwardService` 라고 하면
수상 목록을 읽어올 뿐만 아니라
부문창시상, 인기글 상을 create 하는 역할도 해야한다고 생각했습니다.

그래서 처음에는 이렇게 설계를 했었습니다.
![image](https://user-images.githubusercontent.com/28949340/95555746-39001b80-0a4d-11eb-8dcb-17d192186870.png)

(설계라고 말하고 아무생각없이 구현했다고 읽습니다)

그런데 보시면 순환참조 문제가 생깁니다.
(코드를 막 짜고나니 테스트가 터져서 알게되었습니다)

순환참조 문제를 해결하려면
`AwardService` 에는 부문 상을 만드는 기능이 없어야 했습니다.

그래서 이렇게 수정하게 되었습니다.
(현재버전)
![image](https://user-images.githubusercontent.com/28949340/95555794-4cab8200-0a4d-11eb-8d36-e8ddaaabdf7d.png)

`AwardService` 가 아예 상을 읽기만 하도록
`AwardReadService` 로 재탄생(...)시키게 된 것입니다...

yelimkim98 replied:
코멘트를 읽고나니
`AwardService` 를 `AwardReadService` 라고 할거면
`AwardController` 도 `AwardReadController` 가 되어야 한다는 생각이 들었습니다!

그래서 이 부분을 수정해보았습니다!

aegis1920 replied:
개인적으로는 아예 `SectorService`와 `PopularPostAwardService`에 있는 내용을 모두 `AwardService`에 옮겨도 될 것 같다고 생각이 드네요!
그리고 `AwardReadService` 가 된다면 읽는 역할만 가지게 될텐데 확장성 측면에서 너무 닫혀있지 않을까요?

+ 친절한 설명 감사합니다ㅎㅎ 덕분에 이해가 잘 됐어요!

yelimkim98 replied:
`SectorService`에 있는 내용을 `AwardService`로 옮기려면

부문 상태를 Pending 상태에서 Approved 상태로 바꾸는 일을
`AwardService`가 하게됩니다..

그렇게 할 순 없으니 `SectorService` 에서 부문 창시 상을 만드는 것을 피할 수 없습니다..


그렇다고 `PopularPostAwardService` 에 있는 내용만 `AwardService` 로 옮기자니
부문 창시상은 `SectorAwardService` 에서 만들고 인기글 창시상은 `AwardService` 에서 만들면
통일성이 없어지는 것 같습니다..

수상 관련 로직은
1. 상을 만들거나
2. 상을 읽거나

이 두가지인데,
`AwardReadService` 가 상을 만드는 기능을 가지면 순환참조가 발생하며 문제가 생겨버리니
확장을 할 수 없는 부분이 아닌가 생각됩니다..

aegis1920 replied:
헉 죄송합니다ㅠㅠ `SectorService`가 아니라 `SectorAwardService`와 `PopularPostAwardService`를 말한 거였어요!!

yelimkim98 replied:
아핫..! 그렇군요!!

그치만 그렇게 했을 때 순환참조가 발생해버립니다...

AwardService 가 부문창시상, 인기글상을 create 한다고 가정해보면
AwardService 에서
모든 부문과 모든 지역에 대해 상을 주는 기능을 갖게됩니다.
이 기능을 구현할 때
모든 부문과 모든 지역을 가져와서 이중 for문을 도는데,
이 때 모든 부문을 가져오기 위해서 SectorService 에 의존합니다.

그런데 SectorService는
pending 상태의 부문을 approved 로 바꾸는 기능이 있습니다.
이 기능이 동작할 때 부문 창시상이 생성되기 때문에
SectorService 가 AwardService 에 의존하게됩니다.

이러면 결국 순환참조 문제가 생겨버립니다..

aegis1920 replied:
Service를 의존하지 않고 Repository를 의존해서 구현한다면 순환참조를 없앨 수 있을 것 같은데 이러면 중복되는 코드가 생기는 문제가 있네요. 개인적으로는 중복되는 코드가 생기더라도 AwardService로 만들어서 수상에 대한 모든 책임을 여기에 만드는 게 좋다고 생각하긴 합니다. 어떻게 생각하시나요? 나중에 데일리 때 의견을 나눠봐도 좋구요 :)

yelimkim98 replied:
PopularPostAwardService가 SectorService에 의존하는 대신 SectorRepository에 의존하도록 변경하여
구조를 개선해 보았습니다!!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/util/AwardNameMaker.java:
```diff
@@ -12,4 +16,88 @@
+        LocalDate endDate,
+        RankingCriteria rankingCriteria) {
+
+        if (rankingCriteria.equals(RankingCriteria.LAST_MONTH)) {
```
`rankingCriteria`에게 메시지를 보내도 될 것 같아요 :)

yelimkim98 replied:
`rankingCriteria.isSame(RankingCriteria.LAST_MONTH)`
이것과
`rankingCriteria.equals(RankingCriteria.LAST_MONTH)`
이것의 차이가 없다고 느껴집니다..!

객체에 메세지를 보내자는 측면에서 지금 코드가 문제 없다고 생각합니다!!

aegis1920 replied:
엇 저는 RankingCriteria에 메서드를 만드는 걸 생각했습니다! `rankingCriteria.isLastMonth()` 처럼요!

yelimkim98 replied:
잘못 이해했었네요..!! 그게 좋아 보이네요!!! 반영하겠습니다 감사합니다!!!!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/util/AwardNameMaker.java:
```diff
@@ -12,4 +16,88 @@
+            return makePopularPostAwardNameByMonth(
+                popularPost, ranking, targetArea, startDate, endDate);
+        }
+        throw new UnsupportedOperationException("월간 수상만 지원합니다.");
```
예외 메시지를 보낼 때 현재 어떤 상태인지도 보내주면 예외를 더 정확히 파악할 수 있을 것 같아요 :)

yelimkim98 replied:
그렇네요!! 수정했습니다!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/util/AwardNameMaker.java:
```diff
@@ -12,4 +16,88 @@
+            return makePopularPostAwardNameByMonth(
+                popularPost, ranking, targetArea, startDate, endDate);
+        }
+        throw new UnsupportedOperationException("월간 수상만 지원합니다.");
```
현재 `GlobalExceptionHandler`에 `UnsupportedOperationException`가 없어요!
추가해야될 것 같네요 :)

yelimkim98 replied:
지금 사실상 AwardNameMaker 는 스케쥴링에서만 사용되고 있고 컨트롤러와는 무관한데 의미가 있을지 모르겠습니다..
확실히 스케쥴러에서 UnsupportedOperationException 에 대한 처리가 일어나지 않고있어서 문제네요!
반영해보겠습니다!

aegis1920 replied:
아 그러네요! 서버에서만 사용되니 사용자에게 보여줄 필요가 없네요! 그렇다면 예외를 던지고 로깅만 해줘도 될 것 같다는 생각도 드는데 어떻게 생각하시나요?

yelimkim98 replied:
handler 에서 캐치 하지않고 예외 던질때 로깅 하자는 뜻 맞나요?! 제대로 이해한게 맞는지 모르겠네요!
**
지금은 일단 handler controller 에서
```
@ExceptionHandler({Exception.class, NotFoundAlgorithmException.class})
handleInternalServerError() 
```
에 `UnsupportedOperationException` 을 살포시 추가하고

스케쥴러에 try-catch 를 추가해두었습니다..!

```
@Scheduled(cron = "0 0 0 1 * *")    // cron : 0초 0시 0분 1일 *월 *요일 (매월 1일 0시 0분 0초)
    public void createPopularPostAwards() {
        try {
            popularPostAwardService.createPopularPostAwards(LocalDateTime.now());
        } catch (Exception e) {
            log.error(e.getMessage());
        }
    }
```

aegis1920 replied:
사용자에게 보여줄 필요가 없기 때문에 `controllerAdvice`에서 잡지 않아도 될 것 같아요 :) 오히려 서버의 에러는 사용자에게 보여주면 안되기도 하고요!
그리고 로깅할 때는 저희가 예측하고 있는 예외라면 로깅 레벨을 INFO로 주셔도 됩니다! + 자세한 로깅을 위해서 e.getMessage가 아니라 e를 주는 게 더 좋아요!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/domain/util/AwardNameMaker.java:
```diff
@@ -12,4 +16,88 @@
+    }
+
+    private static void validateMonthForMonthlyPostAward(LocalDate startDate, LocalDate endDate) {
+        if (startDate.getMonthValue() + 1 != endDate.getMonthValue()) {
```
요기서 `1`에 대한 매직넘버를 상수화를 해주면 그 의미를 좀 더 빠르게 파악할 수 있겠네요 :)

yelimkim98 replied:
확실히 if 조건절 가독성이 안좋네요..!

하지만 매직넘버 `1` 에 대해 도저히 어떤 의미를 부여하기가 어려워서
대신 if 조건절 부분을 메서드로 추출해 보았습니다!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/scheduler/PostAwardCreateScheduler.java:
```diff
@@ -0,0 +1,19 @@
+
+    private PopularPostAwardService popularPostAwardService;
+
+    @Scheduled(cron = "0 0 0 1 * *")    // cron : 0초 0시 0분 1일 *월 *요일 (매월 1일 0시 0분 0초)
```
주석으로 쉽게 이해했어요! 👍

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/scheduler/PostAwardCreateScheduler.java:
```diff
@@ -0,0 +1,19 @@
+@Component
+public class PostAwardCreateScheduler {
+
+    private PopularPostAwardService popularPostAwardService;
```
`final`을 붙여도 될 것 같네요!

aegis1920 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/AreaService.java:
```diff
@@ -24,4 +26,9 @@
+
+    @Transactional(readOnly = true)
+    public List<Area> findAllAreas() {
+        return Collections.unmodifiableList(areaRepository.findAll());
```
혹시 불변으로 해준 이유가 있을까요? 궁금합니다!

yelimkim98 replied:
저는 특별히 불변이 아니어야할 상황을 제외하면 모두 불변을 사용하고자 합니다!

all areas를 검색한 뒤에 이 리스트에 뭔가 다른 element를 붙인다면
그 리스트는 더이상 all areas 가 아니게 될 수 있습니다.

그런 코드는 읽는 사람도 불편하고
코드 짜는 사람도 실수하게 만드는 코드라고 생각합니다.

프로그래밍할떄 조금이라도 덜 실수하게 해주는 장치인 것입니다!

aegis1920 replied:
좋네요! 특히 find한 경우에 붙이는 건 더 좋은 습관이라고 생각이 들었어요 :) 설명 감사합니다!

aegis1920 commented on front-end/legeno-around-here/src/components/myProfile/PrivacySection.js:
```diff
@@ -57,6 +57,7 @@ export const PrivacySignOutBox = styled.span`
   font-size: 15px;
   display: inline;
   color: #444444;
+  cursor: pointer
```
세심하네요ㅎㅎ 👍

aegis1920 reviewed (changes requested):
히히 수고했어요~ 몇 가지 피드백 남겼으니 확인해주세요 :)

jamie9504 committed:
refactor: PopularityPostCreatorAward에서 PopularPostAward로 변경

yelimkim98 committed:
Merge pull request #1 from jamie9504/feature/#324

refactor: PopularityPostCreatorAward에서 PopularPostAward로 변경

yelimkim98 committed:
refactor: 컨트롤러 이름 변경

상 목록을 읽기만 하고 상주기 기능을 하지 않는 컨트롤러의 이름을
AwardController 에서 AwardReadController 로 변경했다.
참고로
부문 창시상은 SectorService 에서 부문을 create 할 때 주고있고,
자랑글 인기상은 PostAwardSchedular 에서 매월 정기적으로 주고있다.

yelimkim98 committed:
refactor: 정확한 예외 파악을 위한 에러메세지 구체화

yelimkim98 committed:
feat: UnsupportedOperationException GlobalExceptionHanlder에 추가

yelimkim98 committed:
feat: 스케쥴러 exception catch 추가

yelimkim98 committed:
refactor: final 추가

yelimkim98 committed:
refactor: 의미 이해가 잘 되지 않는 if 조건절을 메서드로 뺌

yelimkim98 committed:
refactor: 객체에 메세지를 보내도록 코드 수정

yelimkim98 committed:
test: 테스트 보강

yelimkim98 committed:
fix: '객체에 메세지를 보내도록 구현' 커밋 되돌리기

yelimkim98 committed:
Revert "refactor: 의미 이해가 잘 되지 않는 if 조건절을 메서드로 뺌"

This reverts commit 924a9a3c10e8ffda63d6e2d95fb851701f69dd7d.

yelimkim98 committed:
Revert "test: 테스트 보강"

This reverts commit 04582f7f0f159d15eed41f504eb11ca6268e479d.

yelimkim98 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#324

 Conflicts:
	back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/service/SectorService.java

yelimkim98 committed:
test: RankingCriteria 테스트케이스 보강 및 이해하기 어려운 도메인에 주석 추가

yelimkim98 committed:
refactor: AwardService 구조 개선 및 AwardNameMaker testcase 보강

yelimkim98 committed:
refactor: 로깅 방식 수정

yelimkim98 committed:
test: 자코코 테스트 커버리지 통과를 위한 PopularPostAwardTest 작성

aegis1920 reviewed (approved):
피드백을 잘 반영해주셨네요! 수고하셨습니다!
